File: app/src/main/java/com/everobo/dooba/core/store/Store.java
package com.everobo.dooba.core.store;

import com.squareup.otto.Bus;
import com.everobo.dooba.core.aciton.Action;


public abstract class Store<T extends Action> {
  private static final Bus bus = new Bus();

  public Store() {
  }

  public void register(final Object view) {
    this.bus.register(view);
  }

  public void unregister(final Object view) {
    this.bus.unregister(view);
  }

  protected void emitStoreChange() {
    this.bus.post(changeEvent());
  }

  public abstract StoreChangeEvent changeEvent();

  public abstract void onAction(T action);

  public static class StoreChangeEvent {
  }
}

File: app/src/main/java/com/everobo/dooba/core/store/Store.kt
package com.everobo.dooba.core.store

import com.squareup.otto.Bus
import com.everobo.dooba.core.aciton.Action


abstract class Store<in T : Action<*>> {

    fun register(view: Any) {
        this.bus.register(view)
    }

    fun unregister(view: Any) {
        this.bus.unregister(view)
    }

    protected fun emitStoreChange() {
        this.bus.post(changeEvent())
    }

    abstract fun changeEvent(): StoreChangeEvent

    abstract fun onAction(action: T)

    class StoreChangeEvent

    private val bus = Bus()
}
